src/cli/self_update/shell.rs

```diff
@@ -255,6 +256,50 @@ impl UnixShell for Fish {
     }
 }
 
+struct Nu;
+
+impl UnixShell for Nu {
+    fn does_exist(&self, process: &Process) -> bool {
+        // nu has to either be the shell or be callable for nu setup.
+        matches!(process.var("SHELL"), Ok(sh) if sh.contains("nu"))
+            || utils::find_cmd(&["nu"], process).is_some()
+    }
+
+    fn rcfiles(&self, process: &Process) -> Vec<PathBuf> {
+        process
+            .var("XDG_CONFIG_HOME")
+            .ok()
+            .and(process.home_dir())
+            .iter()
+            .flat_map(|path| {
+                let mut p0 = path.clone();
+                let mut p1 = path.clone();
+                p0.push("nushell/env.nu");
+                p1.push("nushell/config.nu");
+                [p0, p1]
+            })
+            .collect()
```

Comment:
This doesn't seem like a clear formulation of what the logic is supposed to be. Maybe something like this?

```rust
let mut paths = Vec::new();
if let Ok(config) = process.var("XDG_CONFIG_HOME") {
    let mut path = PathBuf::from(config);
    path.push("nushell/env.nu");
    paths.push(path);
}

if let Some(home) = process.home_dir() {
    home.push("nushell/config.nu");
    paths.push(path);
}

paths
```

Reply:
Nushell will try to look for `XDG_CONFIG_HOME` firstly. If it doesn't existed, Nushell will go to `$HOME/.config` instead. Then `env.nu` and `config.nu` are in the same config dir. 

So the logic should be
```rust
let config_dir = process
            .var("XDG_CONFIG_HOME")
            .ok()
            .and(process.home_dir().map(|dir| dir.join("/.config")));

config_dir.map_or(vec![], |dir| {
  vec![dir.join("env.nu"), dir.join("config.nu")]
})
```

I will fix it soon. BTW, should I just add a new commit in this branch?

Reply:
That code checks if the `XDG_CONFIG_HOME` variable is defined but not whether the directory exists, is that what you meant? (I still think this code sample isn't very easy to follow due to overuse of combinators.)

Ideally please squash additional changes into your existing commit.

Reply:
> Nushell will try to look for `XDG_CONFIG_HOME` firstly. If it doesn't existed, Nushell will go to `$HOME/.config` instead. Then `env.nu` and `config.nu` are in the same config dir.
> 
> So the logic should be
> 
> ```rust
> let config_dir = process
>             .var("XDG_CONFIG_HOME")
>             .ok()
>             .and(process.home_dir().map(|dir| dir.join("/.config")));
> 
> config_dir.map_or(vec![], |dir| {
>   vec![dir.join("env.nu"), dir.join("config.nu")]
> })
> ```
> 
> I will fix it soon. BTW, should I just add a new commit in this branch?

@LittleJianCH That is to say, I believe, the resolution logic is the same logic as `fish`:

https://github.com/rust-lang/rustup/blob/c8eddcb4c1dc9aea8067fa73179ce1791aa24d1d/src/cli/self_update/shell.rs#L221-L233